registry/spec.md

```diff
@@ -1491,71 +1632,1350 @@ An attempt to delete all versions MUST generate an error.
+
+The registry is designed to be extensible to support any structured data
+encoding and related schemas for message or event payloads. The [Schema
+Formats](#schema-formats) section describes the meta-schema for JSON Schema, XML
+Schema, Apache Avro schema, Protobuf schema.
+
+### File format
+
+A CloudEvents Registry can be implemented using the Registry API or with plain
+text files.
+
+When using the file-based model, files with the extension `.cereg` use JSON
+encoding. Files with the extension `.cereg.yaml` or `.cereg.yml` use YAML
+encoding. The formal JSON schema for the file format is defined in the
+[CloudEvents Registry Document Schema](#cloudevents-registry-document-schema),
+which implements the Registry format and the CloudEvents Registry format.
+
+The media-type for the file format is `application/cloudevents-registry+json` for the
+JSON encoding and `application/cloudevents-registry+yaml` for the YAML encoding.
+
+The JSON schema identifier is `https://cloudevents.io/schemas/registry` and the
+`specversion` property indicates the version of this specification that the
+elements of the file conform to.
+
+A CloudEvents Registry file MUST contain a single JSON object or YAML document.
+The object declares the roots of the three sub-registries, which are either
+embedded or referenced. Any of the three sub-registries MAY be omitted.
+
+``` meta
+{
+   "$schema": "https://cloudevents.io/schemas/registry",
+   "specversion": "0.4-wip",
+   "endpoints": { ... } | "endpointsUrl": "URL",
+   "definitiongroups": { ... } | "definitiongroupsUrl": "URL"
+   "schemagroups" : { ... } | "schemagroupsUrl": "URL",
+}
+```
+
+While the file structure leads with endpoints followed by definition groups and
+then schema groups by convention, the order of the sub-registries is not significant.
+
+### Schema Registry
+
+The schema registry is a metadata store for organizing data schemas of any kind.
+
+The Registry API extension model of the Schema Registry is:
+
+``` JSON
+{
+  "groups": [
+    {
+      "singular": "schemagroup",
+      "plural": "schemagroups",
+      "schema": "TBD",
+      "resources": [
+        {
+          "singular": "schema",
+          "plural": "schemas",
+          "versions": -1,
+        }
+      ]
+    }
+  ]
+}
+```
+
+#### Group: schemagroups
+
+The group (GROUP) name for the Schema Registry is `schemagroups`. The group does
+not have any specific extension attributes.
+
+A schema group is a collection of schemas that are related to each other in some
+application-defined way. A schema group does not impose an restrictions on the
+contained schenmas, meaning that a schema group can contain schemas of different
+formats. Every schema MUST reside inside a schema group.
+
+Example:
+
+``` meta
+{
+  "schemagroups": {
+    "com.example.schemas": {
+      "type": "schemagroup",
+      "id": "com.example.schemas",
+      "schemas": {
+        ...
+      }
+    }
+  }
+}
+```
+
+#### Resource: schemas
```

Comment:
let's check that we're consistent on whether things are plural here or not. It feels like it should either be:
Resource: schema
or
Resources: schemas
I prefer the singular version because the 2nd on feels like "schemas" is defining a group not an singular resource type.